BinaryFormatter Migration Guide (#41564)
dotnet/docs: docs/standard/serialization/binaryformatter-migration-guide/wpf-applications.md

Reviewer: @rchauhan18 - I applied my own feedback directly to this page, but please feel free to refine further as needed. I applied the same type of edits to the wpf-applications page too.

/cc @terrajobst @gewarren

Reviewer 2: Thanks @jeffhandley

---
title: "BinaryFormatter migration guide: WPF applications"
description: "Learn about the effects of the deprecation and removal of BinaryFormatter from .NET on Windows Forms and how to migrate."
ms.date: 8/05/2024
no-loc: [BinaryFormatter, WPF]
helpviewer_keywords:
  - "BinaryFormatter"
  - "WPF"
---

> # WPF Migration Guide – Binary Formatter 

Reviewer 3: Consider writing out the entire name for the first time
```suggestion
# Windows Presentation Foundation (WPF) Migration Guide – BinaryFormatter 
```

 
## Binary Formatter Removal 

> In .NET 9.0, the entire `Binaryformatter` infrastructure will be removed from the product and moved as a separate Nuget  package. For more information about the risks `Binaryformatter` poses and our decision on its removal, see [BinaryFormatter migration guide](index.md). 

Reviewer 3: ```suggestion
In .NET 9.0, the entire `Binaryformatter` infrastructure will be removed from the product and moved as a separate Nuget  package. For more information about the risks `BinaryFormatter` poses and the reason for its removal, see [BinaryFormatter migration guide](index.md).
```

>  
> ## How Binary Formatter affects WPF 

Reviewer 3: ```suggestion
## How BinaryFormatter affects WPF
```

> WPF uses `Binaryformatter` to serialize/deserialize the data for scenarios such clipboard, drag/drop and store data for Avalon Binding in Journal.  

Reviewer 3: ```suggestion
WPF uses `Binaryformatter` to serialize/deserialize the data for scenarios such clipboard, drag-and-drop and store data for Avalon Binding in Journal.  
```

> WPF uses `Binaryformatter` to serialize/deserialize the data for scenarios such clipboard, drag/drop and store data for Avalon Binding in Journal.  

Reviewer 4: I like to think I have a decent understanding of our tech and I have no idea what that refers to :-)

WPF has and is still currently using `Binaryformatter` code internally for all the above scenarios, however we have taken some measures to mitigate the usage of `Binaryformatter`.  

We have added an implementation based on WinForms to handle common primitive types, including primitive arrays and lists. Usage of these types in the mentioned scenarios will continue to work without using `Binaryformatter`, similarly as before. 

Primitive types include: 
- `bool` 
- `byte` 
- `char` 
- `decimal` 
- `double` 
- `int` 
- `sbyte` 
- `float` 
- `TimeSpan` 
- `DateTime` 
- `uint` 
- `string` 
- `nint` 
- `nuint` 
- `long`
- `ulong`
- `short`
- `ushort` 

Additional types include: 
- `PointF` 
- `RectangleF`  
  
## Where Binary Formatter is used in WPF 

> ### Clipboard  

Reviewer 3: I had started adding clipboard section in `winforms-wpf-ole-guidance.md`. My understanding is that WinForms and WPF will hit BinaryFormatter in the same way. If you could review and let me know if it is accurate for WPF that would be great. Then, we can update this section to point to that doc instead.

Reviewer 2: Thanks @lonitra for adding the common OLE scenarios for WPF and WinForms.
The changes are LGTM.
I have updated the section to point to the `winforms-wpf-ole-guidance.md`

In WPF application if you are using a clipboard to copy or paste the data into the application or Set/Get data from a DataObject or using drag/drop. 

> If the data format ([DataFromats](https://learn.microsoft.com/en-us/dotnet/api/system.windows.dataformats?view=windowsdesktop-8.0)) is part of the following list, it will not require BinaryFormatter to handle: 

Reviewer 5: ```suggestion
If the [data format](/dotnet/api/system.windows.dataformats) is included in the following list, it doesn't require using `BinaryFormatter` to handle:

```

- `Stream` 
- `Text`, `Rtf`, `OemText`, `CommaSeparatedValue` 
- `UnicodeText` 
- `FileDrop` 
- `FileName` 
- `BitmapSource` 

> Apart from these formats, if the type is serializable and not included in the above list of types handled by the new implementation, `Binaryformatter` will be used for handling. 

Reviewer 5: ```suggestion
However, if the type is serializable and not included in the previous list of types that are handled by the new implementation, `BinaryFormatter` is used. 
```

> In scenarios where `Binaryformatter` is utilized, its removal will result in developers encountering an exception indicating that `Binaryformatter` has been removed. 

Reviewer 5: ```suggestion
In scenarios where `BinaryFormatter` is used, its removal will result in a run-time exception that indicates that `BinaryFormatter` has been removed. 
```

You can refer to the function where we have used `Binaryformatter` as fallback to read/save object to handle:  [SaveObjectToHandle](https://github.com/dotnet/wpf/blob/0354a597996adae43b12efc72bd705f76d4ba497/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/dataobject.cs#L1677) and  [ReadObjectFromHandle](https://github.com/dotnet/wpf/blob/0354a597996adae43b12efc72bd705f76d4ba497/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/dataobject.cs#L3051) 

> Additionally, if you encounter an exception stating, 'Data stored in clipboard is invalid,' it may be because `Binaryformatter` was used to save the data. 

Reviewer 5: ```suggestion
Additionally, if you encounter an exception stating, 'Data stored in clipboard is invalid,' it may be because `Binaryformatter` was used to save the data. 
```
```suggestion
Additionally, if you encounter an exception that states "Data stored in clipboard is invalid," it might be because `BinaryFormatter` was used to save the data. 
```

Ref: [DataObject.cs](https://github.com/dotnet/wpf/blob/4e977f5fe8c73094ee5826dbfa7a0f37c3bf0e33/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/dataobject.cs) 

 

### Drag/Drop 

In drag/drop any UI element or Content element can participate, for an event is raised which is defined in [DragDrop.cs](https://github.com/dotnet/wpf/blob/0354a597996adae43b12efc72bd705f76d4ba497/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/DragDrop.cs). In the Drag/Drop operation a DataObject is used to store the data and to Get/Set data we call `GetData()`/`SetData()`. In this scenario if the type involved is not intrinsically handled during serialization/deserialization, `Binaryformatter` is used. 

With `Binaryformatter` removed, developers will now encounter an exception indicating that `Binaryformatter` has been removed. 
	 

### Journaling 

In the case when we need to store or load a state while managing the navigation history in WPF. 

> To load/save we call [LoadSubStreams](https://github.com/dotnet/wpf/blob/0354a597996adae43b12efc72bd705f76d4ba497/src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/DataStreams.cs#L244)/ [SaveSubStreams](https://github.com/dotnet/wpf/blob/0354a597996adae43b12efc72bd705f76d4ba497/src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/DataStreams.cs#L86) of DataStream class. If the element used in not part of know type handled by the new implementation, it will use `Binaryformatter`. 

Reviewer 3: Is there some sort of user action that triggers the call to load/save? That might be good to call out if so.

Reviewer 2: User can call save/load just by navigating to/back/forward, as at each step SaveState function of JournalEntry is called which further trigger SaveSubStreams/LoadSubStream.
I used this scenario for testing, there could be other scenarios too.

---
title: "BinaryFormatter migration guide: WPF applications"
description: "Learn about the effects of the deprecation and removal of BinaryFormatter from .NET on WPF and how to migrate."
ms.date: 8/05/2024
no-loc: [BinaryFormatter, WPF]
helpviewer_keywords:
  - "BinaryFormatter"
  - "WPF"
---

# WPF migration guide for BinaryFormatter

## BinaryFormatter removal

Starting with .NET 9, `BinaryFormatter` is no longer supported due to its known [security risks](../binaryformatter-security-guide.md) and its APIs always throw an exception for all project types, including WPF apps. For more information about the risks BinaryFormatter poses and the decision on its removal, see the [BinaryFormatter migration guide](index.md).

With BinaryFormatter’s removal, it's expected that many WPF applications will be impacted, and you'll need to take action to complete your migration to .NET 9 or a later version.

## How BinaryFormatter affects WPF

> Prior to .NET 9, Windows Presentation Foundation (WPF) used `BinaryFormatter` to serialize and deserialize data for scenarios such clipboard, drag-and-drop, and to store data for Avalon Binding in Journal. Starting with .NET 9, WPF and Windows Forms use a subset of the `BinaryFormatter` implementation internally for these scenarios. While BinaryFormatter's risks cannot be addressed in general-purpose serialization/deserialization, measures have been taken to mitigate the risks in these very specific use cases with a known set of types. A fall-back to `BinaryFormatter` is still in place for unknown or unsupported types, which will throw exceptions unless migration steps are taken in the application.

Reviewer: I'm unfamiliar with Avalon Binding in Journal. Is 'Avalon' a well-known part of this feature name and/or is there a doc we should link to for this?

Reviewer 4: Many people know that Avalon was the codename for WPF (that's why there is Avalonia) but it's not used in our docs. Unless we can point to something specific, I'm in favor of deleting the reference to 'Avalon Binding in Journal'.

Reviewer 2: Thanks for pointing it out.
I wanted to convey "Holds the data for Avalon BindProducts in the journal",  
I have updated it.
